skbeam/core/correlation.py

```diff
@@ -911,93 +909,130 @@ def __init__(self, shape, mask=None, normalization=None,
+            ppii[index_start:index_stop] -= shapes[i, 0]
+            ppjj[index_start:index_stop] -= shapes[i, 2]
+        # now save to object
+        self.ppii = ppii
+        self.ppjj = ppjj
+        self.pi = pi
+        self.pj = pj
+
+        # redefine shapes to be the shapes of the subregions
+        # make shapes be for regions
+        shapes = np.array(1 + (np.diff(shapes, axis=-1)[:, [0, 2]]))
+        self.shapes = shapes
+        # We now have two sets of positions of the subregions (pi,pj) in
+        # subregion and (pii,pjj) in images. pos is a pointer such that
+        # (pos[i]:pos[i+1]) is the indices in the position arrays of subregion
+        # i.
 
         # Making a list of arrays holding the masks for each id. Ideally, mask
         # is binary so this is one element to quickly index original images
-        self.pxlsts = list()
         self.submasks = list()
-        # to quickly index the sub images
-        self.subpxlsts = list()
-        # the temporary images (double the size for the cross correlation)
-        self.tmpimgs = list()
-        self.tmpimgs2 = list()
         self.centers = list()
-        self.shapes = list()  # the shapes of each correlation
         # the positions of each axes of each correlation
         self.positions = list()
-
-        self.ids = np.sort(np.unique(mask))
-        # remove the zero since we ignore, but only if it is there (sometimes
-        # may not be)
-        if self.ids[0] == 0:
-            self.ids = self.ids[1:]
-
-        self.nids = len(self.ids)
         self.maskcorrs = list()
         # regions where the correlations are not zero
         self.pxlst_maskcorrs = list()
 
         # basically saving bunch of mask related stuff like indexing etc, just
         # to save some time when actually computing the cross correlations
-        for idno in self.ids:
-            masktmp = (mask == idno)
-            self.pxlsts.append(np.where(masktmp.ravel() == 1)[0])
-
-            # this could be replaced by skimage cropping and padding
-            submasktmp = _crop_from_mask(masktmp)
-
-            if self.wrap is False:
-                submask = _expand_image(submasktmp)
-
-            tmpimg = np.zeros_like(submask)
-
+        for i in range(self.nids):
+            submask = np.zeros(self.shapes[i, :])
+            index_start, index_stop = idpos[i], idpos[i+1]
+            ppiis = ppii[index_start:index_stop]
+            ppjjs = ppjj[index_start:index_stop]
+            submask[ppiis, ppjjs] = 1
             self.submasks.append(submask)
-            self.subpxlsts.append(np.where(submask.ravel() == 1)[0])
-            self.tmpimgs.append(tmpimg)
-            # make sure it's a copy and not a ref
-            self.tmpimgs2.append(tmpimg.copy())
+
             maskcorr = _cross_corr(submask)
-            # quick fix for finite numbers should be integer so
             # choose some small value to threshold
             maskcorr *= maskcorr > .5
```

Comment:
I would leave it. The FFT introduces finite numbers even at zero values (as opposed to brute force convolution). But they're very small. Since we know results should be at least 1 or greater (since the submask is essentially 1's and 0's).

However, perhaps a better way to do this to round to the nearest integer. You can't just typecast to int though, I think. Reason is beacuse `maskcorr.astype(int)` for values like 2.99999 would become 2 when they should be 3 etc.)
Maybe it's something to try for a later pull request?